UPGRADE-4.0.md

```diff
@@ -26,3 +26,8 @@ If you have created a custom `AbstractBlockService` you must now implement the n
 Blocks are using `twig` service to render templates, so if you already overrode a constructor of a custom `AbstractBlockService`, you must update it.
 
 Also, both arguments of `AbstractBlockService` constructor became required.
+
+## Closed API
+
+Many classes have been made final, meaning you can no longer extend them.
```

Comment:
I know it could be hard, but I think we should make explicit the list of classes which are declared as final.

Reply:
Just copy & pasted this:
https://github.com/sonata-project/exporter/blob/2.x/UPGRADE-2.0.md

Reply:
I'd like to improve that too.
Please, let me know if you need some help with the list and I'll try to provide what I think it should contain based on your changes.
IMO, an explicit upgrade path is much easier to follow and gives a much better DX.

Reply:
I'll have a look at this tomorrow

Reply:
I think this is the list:
```
  * `Sonata\BlockBundle\Annotation\Block`;
  * `Sonata\BlockBundle\Block\BlockContext`;
  * `Sonata\BlockBundle\Block\BlockContextManager`;
  * `Sonata\BlockBundle\Block\BlockLoaderChain`;
  * `Sonata\BlockBundle\Block\BlockRenderer`;
  * `Sonata\BlockBundle\Block\BlockServiceManager`;
  * `Sonata\BlockBundle\Block\Loader\ServiceLoader`;
  * `Sonata\BlockBundle\Block\Service\ContainerBlockService`;
  * `Sonata\BlockBundle\Block\Service\EmptyBlockService`;
  * `Sonata\BlockBundle\Block\Service\MenuBlockService`;
  * `Sonata\BlockBundle\Block\Service\RssBlockService`;
  * `Sonata\BlockBundle\Block\Service\TemplateBlockService`;
  * `Sonata\BlockBundle\Block\Service\TextBlockService`;
  * `Sonata\BlockBundle\Cache\HttpCacheHandler`;
  * `Sonata\BlockBundle\Cache\NoopHttpCacheHandler`;
  * `Sonata\BlockBundle\Command\DebugBlocksCommand`;
  * `Sonata\BlockBundle\DependencyInjection\Compiler\GlobalVariablesCompilerPass`;
  * `Sonata\BlockBundle\DependencyInjection\Compiler\TweakCompilerPass`;
  * `Sonata\BlockBundle\DependencyInjection\Configuration`;
  * `Sonata\BlockBundle\DependencyInjection\SonataBlockExtension`;
  * `Sonata\BlockBundle\Event\BlockEvent`;
  * `Sonata\BlockBundle\Exception\BlockNotFoundException`;
  * `Sonata\BlockBundle\Exception\BlockOptionsException`;
  * `Sonata\BlockBundle\Exception\Filter\DebugOnlyFilter`;
  * `Sonata\BlockBundle\Exception\Filter\IgnoreClassFilter`;
  * `Sonata\BlockBundle\Exception\Filter\KeepAllFilter`;
  * `Sonata\BlockBundle\Exception\Filter\KeepNoneFilter`;
  * `Sonata\BlockBundle\Exception\Renderer\InlineDebugRenderer`;
  * `Sonata\BlockBundle\Exception\Renderer\InlineRenderer`;
  * `Sonata\BlockBundle\Exception\Renderer\MonkeyThrowRenderer`;
  * `Sonata\BlockBundle\Exception\Strategy\StrategyManager`;
  * `Sonata\BlockBundle\Form\Type\ContainerTemplateType`;
  * `Sonata\BlockBundle\Form\Type\ServiceListType`;
  * `Sonata\BlockBundle\Meta\Metadata`;
  * `Sonata\BlockBundle\Model\EmptyBlock`;
  * `Sonata\BlockBundle\Profiler\DataCollector\BlockDataCollector`;
  * `Sonata\BlockBundle\SonataBlockBundle`;
  * `Sonata\BlockBundle\Twig\Extension\BlockExtension`;
  * `Sonata\BlockBundle\Twig\GlobalVariables`;
  * `Sonata\BlockBundle\Util\RecursiveBlockIterator`;
  * `Sonata\BlockBundle\Util\RecursiveBlockIteratorIterator`;
  * `Sonata\BlockBundle\Tests\Block\BlockContextManagerTest`;
  * `Sonata\BlockBundle\Tests\Block\BlockExecutionContextTest`;
  * `Sonata\BlockBundle\Tests\Block\BlockLoaderChainTest`;
  * `Sonata\BlockBundle\Tests\Block\BlockRendererTest`;
  * `Sonata\BlockBundle\Tests\Block\BlockServiceManagerTest`;
  * `Sonata\BlockBundle\Tests\Block\Service\MenuBlockServiceTest`;
  * `Sonata\BlockBundle\Tests\Block\Service\RssBlockServiceTest`;
  * `Sonata\BlockBundle\Tests\Block\Service\TextBlockServiceTest`;
  * `Sonata\BlockBundle\Tests\Cache\HttpCacheHandlerTest`;
  * `Sonata\BlockBundle\Tests\Cache\NoopHttpCacheHandlerTest`;
  * `Sonata\BlockBundle\Tests\Exception\Renderer\InlineDebugRendererTest`;
  * `Sonata\BlockBundle\Tests\Exception\Renderer\InlineRendererTest`;
  * `Sonata\BlockBundle\Tests\Exception\Renderer\MonkeyThrowRendererTest`;
  * `Sonata\BlockBundle\Tests\Exception\Strategy\StrategyManagerTest`;
  * `Sonata\BlockBundle\Tests\Form\Type\ServiceListTypeTest`;
  * `Sonata\BlockBundle\Tests\Model\BlockTest`;
  * `Sonata\BlockBundle\Tests\Naming\ConvertFromFqcnTest`;
  * `Sonata\BlockBundle\Tests\Templating\Helper\BlockHelperTest
```